src/bit-fields/src/bit_range.rs

```diff
@@ -0,0 +1,1045 @@
+                assert_eq!(BitRangeMut::<_,0,6>(&mut x).get_max(),", stringify!($y), "::new(63).unwrap());
+                assert_eq!(BitRangeMut::<_,0,7>(&mut x).get_max(),", stringify!($y), "::new(127).unwrap());
+                assert_eq!(BitRangeMut::<_,0,8>(&mut x).get_max(),", stringify!($y), "::new(255).unwrap());
+                ```
+            ")]
+            #[must_use]
+            #[inline]
+            pub const fn get_max(&self) -> $y {
+                Self::MAX
+            }
+            #[doc = concat!("
+                Convenience alias for [`Self::MASK`].
+                
+                ```
+                use bit_fields::BitRangeMut;
+                use std::num::", stringify!($y), ";
+                let mut x = ", stringify!($y), "::new(1).unwrap();
+
+                assert_eq!(BitRangeMut::<_,0,1>(&mut x).get_mask(),", stringify!($y), "::new(1).unwrap());
+                assert_eq!(BitRangeMut::<_,1,3>(&mut x).get_mask(),", stringify!($y), "::new(6).unwrap());
+                assert_eq!(BitRangeMut::<_,3,7>(&mut x).get_mask(),", stringify!($y), "::new(120).unwrap());
+                assert_eq!(BitRangeMut::<_,7,8>(&mut x).get_mask(),", stringify!($y), "::new(128).unwrap());
+                ```
+            ")]
+            #[must_use]
+            #[inline]
+            pub const fn get_mask(&self) -> $y {
+                Self::MASK
+            }
+        }
+    };
+}
+
+bit_mut_range!(u128, NonZeroU128, mask_u128, max_u128);
+bit_mut_range!(u64, NonZeroU64, mask_u64, max_u64);
+bit_mut_range!(u32, NonZeroU32, mask_u32, max_u32);
+bit_mut_range!(u16, NonZeroU16, mask_u16, max_u16);
+bit_mut_range!(u8, NonZeroU8, mask_u8, max_u8);
+
+/// Macro for defining `From` implementations on `BitRange` and `BitRangeMut`.
+macro_rules! bit_range_from {
+    ($x:ty, $y:ty, $bit_range: ident) => {
+        // `START < 8 * size_of::<$x>()` is always true so the right shift will not panic.
+        #[allow(clippy::integer_arithmetic, clippy::arithmetic_side_effects)]
+        impl<const START: u8, const END: u8> From<&$bit_range<'_, $x, START, END>> for $x {
+            #[inline]
+            fn from(this: &$bit_range<'_, $x, START, END>) -> Self {
+                let a = $bit_range::<'_, $x, START, END>::MASK & *this.0;
+                a >> START
+            }
+        }
+        // `START < 8 * size_of::<$x>()` is always true so the right shift will not panic.
+        #[allow(clippy::integer_arithmetic, clippy::arithmetic_side_effects)]
+        impl<const START: u8, const END: u8> From<&$bit_range<'_, $y, START, END>> for $y {
+            #[inline]
+            fn from(this: &$bit_range<'_, $y, START, END>) -> Self {
+                let a = $bit_range::<'_, $x, START, END>::MASK & this.0.get();
+                // SAFETY: Always safe.
+                unsafe { <$y>::new_unchecked(a >> START) }
+            }
+        }
```

Comment:
This implementation is unsafe. Consider the following bitfield

```rs
bitfield!(ExampleU8, NonZeroU8 {
    my_bit: 0,
    my_range: 1..7,
});
```

It can have the value `0b0000_0001`, e.g. have `my_bit` set to 1 and `my_range` set to all zeros. Then this `From` impl would construct `NonZeroU8::new_unchecked(0)`. 

The problem here is that having a bitfield with a NonZero backing type does not mean all its ranges are also non-zero. From should return `u8` instead of `NonZeroU8` here

Reply:
Completed a rework to the functions on bit-fields with `NonZero` backing types that should fix this.